Channel offsets causes SVT readout simulation to crash

# Issues

* When retrieving the hit queues (see SVTReadoutDriver::process), the index used maps directly to the channel ID of a hit.  For the thin sensors, this is no longer valid because of the channel offset and leads to the following error 


```
java.lang.ArrayIndexOutOfBoundsException: Index 510 out of bounds for length 510
	at org.hps.readout.svt.SVTReadoutDriver.process(SVTReadoutDriver.java:274)
	at org.lcsim.util.Driver.doProcess(Driver.java:261)
	at org.lcsim.util.Driver.processChildren(Driver.java:271)
	at org.lcsim.util.Driver.process(Driver.java:187)
	at org.lcsim.util.DriverAdapter.recordSupplied(DriverAdapter.java:74)
	at org.freehep.record.loop.DefaultRecordLoop.consumeRecord(DefaultRecordLoop.java:832)
	at org.freehep.record.loop.DefaultRecordLoop.loop(DefaultRecordLoop.java:668)
	at org.freehep.record.loop.DefaultRecordLoop.execute(DefaultRecordLoop.java:566)
	at org.lcsim.util.loop.LCSimLoop.loop(LCSimLoop.java:152)
	at org.lcsim.job.JobControlManager.run(JobControlManager.java:949)
	at org.hps.job.JobManager.main(JobManager.java:33)
org.lcsim.job:SEVERE A fatal error occurred when running the job.
java.lang.RuntimeException: java.lang.ArrayIndexOutOfBoundsException: Index 510 out of bounds for length 510
	at org.lcsim.util.loop.LCSimLoop.handleClientError(LCSimLoop.java:194)
	at org.freehep.record.loop.DefaultRecordLoop.consumeRecord(DefaultRecordLoop.java:834)
	at org.freehep.record.loop.DefaultRecordLoop.loop(DefaultRecordLoop.java:668)
	at org.freehep.record.loop.DefaultRecordLoop.execute(DefaultRecordLoop.java:566)
	at org.lcsim.util.loop.LCSimLoop.loop(LCSimLoop.java:152)
	at org.lcsim.job.JobControlManager.run(JobControlManager.java:949)
	at org.hps.job.JobManager.main(JobManager.java:33)
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 510 out of bounds for length 510
	at org.hps.readout.svt.SVTReadoutDriver.process(SVTReadoutDriver.java:274)
	at org.lcsim.util.Driver.doProcess(Driver.java:261)
	at org.lcsim.util.Driver.processChildren(Driver.java:271)
	at org.lcsim.util.Driver.process(Driver.java:187)
	at org.lcsim.util.DriverAdapter.recordSupplied(DriverAdapter.java:74)
	at org.freehep.record.loop.DefaultRecordLoop.consumeRecord(DefaultRecordLoop.java:832)
	... 5 more
```



* Another issue related to this occurs when processing the hit queues after a trigger signal is received (see SVTReadoutDriver::getOnTriggerData).  Currently, the hit queues are retrieved by index where the max is equal to the total number of channels in a sensor.  The same index is used to retrieve the pedestals and noise from the conditions database and also used to create the channel ID.   For thin sensors, using the index will lead to the retrieval of the wrong condition because of the channel offset.  The channel ID will also be wrong because of this. 

# Solution

Apply a channel offset when hits from a L0/L1 sensor are being processed.

--- a/detector-model/src/main/java/org/lcsim/detector/tracker/silicon/SiStriplets.java
+++ b/detector-model/src/main/java/org/lcsim/detector/tracker/silicon/SiStriplets.java
@@ -48,28 +48,21 @@ public class SiStriplets extends SiPixels {
      * Then strip can be used as cellID in this class to get row/columns/position 
      *
      **/
-
     @Override
     public int getCellID(int row, int column) { 
-        //System.out.println("[ getCellID ]: Row: " + row + " Col: " + column); 
         if (row < 0 || row >= getNCells(0)) return -1; 
         
         if (column < 0 || column >= getNCells(1)) return -1; 
 
         int id = (column == 0) ? row + CHANNEL_MAP_OFFSET : ((column + 1)*getNCells(0) - row); 
        
-        //System.out.println("[ getCellID ]: Original Cell ID: " + id); 
-        //System.out.println("[ getCellID ]: New Cell ID: " + (id + CHANNEL_MAP_OFFSET));
-        //System.out.println("[ getCellID ]: Position: " + getCellPosition(id + CHANNEL_MAP_OFFSET).toString());  
         return id;
     }
 
     
     @Override 
     public int getRowNumber(int cellID) {
-        //System.out.println("[ getRowNumber ]: Cell ID: " + cellID); 
         int row = (getColumnNumber(cellID) == 0) ? (cellID - CHANNEL_MAP_OFFSET): (getColumnNumber(cellID) + 1)*getNCells(0) - cellID;
-        //System.out.println("[ getRowNumber ]: Cell ID: " + cellID + " Before Off Row: " + row); 
        
         return row;      
     }
@@ -77,28 +70,19 @@ public class SiStriplets extends SiPixels {
     @Override
     public int getColumnNumber(int cellID) {
         int col = (int) Math.floor((cellID - CHANNEL_MAP_OFFSET)/getNCells(0));  
-        //System.out.println("[ getColumnNumber ]: Cell ID: " + cellID + " Col: " + col); 
         return col; 
     }
 
     @Override
     public int getRowNumber(Hep3Vector position) {
-        //System.out.println("[ getRowNumber(position) ]: Position: " + position.toString());  
         int row = super.getRowNumber(position);
-        //System.out.println("[ getRowNumber(position) ]: Row: " + row); 
         return row;
     }
 
     public Hep3Vector getStripCenter(int cellID) { 
 
-        //System.out.println("SiStriplets::getStripCenter : Row Offset: " + _row_offset); 
-        //System.out.println("SiStriplets::getStripCenter : Cell ID: " + cellID); 
-        //System.out.println("SiStriplets::getStripCenter : Row: " + getRowNumber(cellID));  
         double u = getRowNumber(cellID)*_row_pitch - _row_offset;   
-        //System.out.println("SiStriplets::getStripCenter : u: " + u);
-
         double v = getColumnNumber(cellID)*_col_pitch - _col_offset; 
-        //System.out.println("SiStriplets::getStripCenter : v: " + v);
 
         return new BasicHep3Vector(u, v, 0.0); 
     }

--- a/tracking/src/main/java/org/hps/readout/svt/SVTReadoutDriver.java
+++ b/tracking/src/main/java/org/hps/readout/svt/SVTReadoutDriver.java
@@ -13,6 +13,7 @@ import org.hps.conditions.svt.SvtTimingConstants;
 import org.lcsim.detector.tracker.silicon.ChargeCarrier;
 import org.lcsim.detector.tracker.silicon.HpsSiSensor;
 import org.lcsim.detector.tracker.silicon.SiSensor;
+import org.lcsim.detector.tracker.silicon.SiStriplets; 
 import org.lcsim.geometry.Detector;
 import org.lcsim.lcio.LCIOConstants;
 import org.lcsim.event.EventHeader;
@@ -250,6 +251,7 @@ public class SVTReadoutDriver extends ReadoutDriver {
     
     @Override
     public void process(EventHeader event) {
+
         super.process(event);
         
         // Generate the truth hits.
@@ -267,6 +269,17 @@ public class SVTReadoutDriver extends ReadoutDriver {
                 // Get the sensor and channel for the truth hit.
                 SiSensor sensor = stripHit.sensor;
                 int channel = stripHit.channel;
+               
+                // This is required to account for the fact that the channel 
+                // number of SiStriplets is shifted by 1. If left uncorrected, 
+                // this leads to an 'out of bounds' error when trying to access
+                // the queue associated with channel 510.
+                // TODO: A cleaner solution is to add a CHANNEL_OFFSET parameter 
+                //       to all instances of readout electrodes and use that to
+                //       shift the channel. For SiStrips, the offset would be
+                //       0.
+                if (sensor.getReadoutElectrodes(ChargeCarrier.HOLE) instanceof SiStriplets)
+                    channel -= 1; 
                 
                 // Queue the hit in the processing queue appropriate
                 // to its sensor and channel.
@@ -420,6 +433,7 @@ public class SVTReadoutDriver extends ReadoutDriver {
                     
                     // Loop over all sensor channels.
                     for(Integer channel : electrodeDataCol.keySet()) {
+
                         // Get the electrode data for this channel.
                         SiElectrodeData electrodeData = electrodeDataCol.get(channel);
                         Set<SimTrackerHit> simHits = electrodeData.getSimulatedHits();
@@ -584,9 +598,22 @@ public class SVTReadoutDriver extends ReadoutDriver {
         for(SiSensor sensor : sensors) {
             // Get the hit queues for the current sensor.
             PriorityQueue<StripHit>[] hitQueues = hitMap.get(sensor);
+
+            // Check if a channel offset needs to be applied.  This will be 
+            // needed to retrieve the proper conditions for queues associated
+            // with SiStriplets.
+            int channelOffset = 0; 
+            if (sensor.getReadoutElectrodes(ChargeCarrier.HOLE) instanceof SiStriplets)
+                channelOffset = 1;  
             
             // Iterate over the hit queue channels.
-            for(int channel = 0; channel < hitQueues.length; channel++) {
+            for(int channel = 0; channel < hitQueues.length; channel++) { 
+               
+                // Use this to retrieve the correct conditions from the 
+                // conditions database.  This accounts for the shift in the 
+                // channel map in the thin sensors.
+                int conditionsChannel = channel + channelOffset; 
+                
                 // Unless noise should be added, there is nothing to
                 // process on an empty hit queue. Skip it.
                 if(!addNoise && (hitQueues[channel] == null || hitQueues[channel].isEmpty())) {
@@ -598,12 +625,12 @@ public class SVTReadoutDriver extends ReadoutDriver {
                 // pedestal values.
                 double[] signal = new double[6];
                 for(int sampleN = 0; sampleN < 6; sampleN++) {
-                    signal[sampleN] = ((HpsSiSensor) sensor).getPedestal(channel, sampleN);
+                    signal[sampleN] = ((HpsSiSensor) sensor).getPedestal(conditionsChannel, sampleN);
                 }
                 
                 // If noise should be added, do so.
                 if(addNoise) {
-                    addNoise(sensor, channel, signal);
+                    addNoise(sensor, conditionsChannel, signal);
                 }
                 
                 // Create a list to store truth SVT hits.
@@ -628,7 +655,7 @@ public class SVTReadoutDriver extends ReadoutDriver {
                             double signalAtTime = hit.amplitude * shape.getAmplitudePeakNorm(sampleTime - hit.time);
                             totalContrib += signalAtTime;
                             signal[sampleN] += signalAtTime;
-                            meanNoise += ((HpsSiSensor) sensor).getNoise(channel, sampleN);
+                            meanNoise += ((HpsSiSensor) sensor).getNoise(conditionsChannel, sampleN);
                             
                             signalBuffer.append(signalAtTime + " (" + sampleTime + ")");
                             if(sampleN != 5) {
@@ -658,7 +685,9 @@ public class SVTReadoutDriver extends ReadoutDriver {
                 }
                 
                 // Get the proper channel ID.
-                long channel_id = ((HpsSiSensor) sensor).makeChannelID(channel);
+                // For the thin sensors, the channel number needs to be offset
+                // by 1. 
+                long channel_id = ((HpsSiSensor) sensor).makeChannelID(conditionsChannel);
                 
                 // Create a new tracker hit.
                 RawTrackerHit hit = new BaseRawTrackerHit(0, channel_id, samples, simHits, sensor);

